Added option for encrypting secrets to etcd

woopstar opened the pull request:
This PR will add the experimental option for encrypting secrets in ETCD

Based on https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/ and https://www.twistlock.com/2017/08/02/kubernetes-secrets-encryption/

woopstar commented:
I'm considering to do a `kubectl get secrets --all-namespaces -o json | kubectl replace -f -` as it will encrypt all secrets created.

But not sure where to put that. I'm thinking of a handler, at notifying it at the "right" time.

gopinatht commented:
Needs the `25` changed to `32`  in the line ```encrypt_token: "{{ lookup('password', 'credentials/encrypt_token length=25 chars=ascii_letters,digits')```.  Once that fix is in, I can LGMT this (do I need to be on some approver list for it to matter?)

woopstar commented:
Disregard my comment about encrypting secrets. I believe at the place I've added the task will cause this option to be set before the api-server is initially started, so all tokens will be encrypted from the start

woopstar commented:
@ant31 Can you CI check this?

chapsuk commented:
ci check this

chapsuk commented on roles/kubernetes/master/tasks/encrypt-at-rest.yml:
```diff
@@ -0,0 +1,7 @@
+---
+- name: Write secrets for encrypting secret data at rest
+  template:
+    src: secrets.yaml.j2
```
How about rename this file to `secrets_encryption.yaml.j2` ?

woopstar replied:
Sure. Let me fix that.

chapsuk commented on roles/kubernetes/master/templates/kubeadm-config.yaml.j2:
```diff
@@ -48,6 +48,9 @@ apiServerExtraArgs:
 {%   endif %}
+{% endif %}
+{% if encrypt_secret_data|default(false) %}
+  experimental-encryption-provider-config: /etc/kubernetes/secrets_encryption.yaml
```
it should be `{{ kube_config_dir }}/secrets_encryption.yaml`

chapsuk commented on roles/kubernetes/master/templates/secrets_encryption.yaml.j2:
```diff
@@ -0,0 +1,11 @@
+    - {{ encryption_algorithm }}:
+        keys:
+        - name: key
+          secret: {{ encrypt_token }}
```
it should be `base64` encode from encryption token, not a plain token

woopstar replied:
You are right! Fixed

gopinatht commented:
Quick question. Would this work in API Server HA mode? i.e., the same encryption token should be used for all API server instances.

gopinatht commented:
It looks like a new encryption token will be generated for each node that this play will run in. That means if we have K8S API server HA, each API server instance will have a different password (encryption token). Ansible says we should use vault if we do not want per node password: http://docs.ansible.com/ansible/latest/playbooks_vault.html

gopinatht commented:
Got clarification from @woopstar in slack about the generation of password. I am OK with this implementation. LGMT.

mattymo commented on roles/kubernetes/master/tasks/main.yml:
```diff
@@ -12,6 +12,9 @@
   when: kube_basic_auth|default(true)
 
+- import_tasks: encrypt-at-rest.yml
+  when: encrypt_secret_data|default(false)
```
There's no need for |default(false) since there is a default set in the role.

mattymo commented on roles/kubernetes/master/defaults/main.yml:
```diff
@@ -88,3 +88,8 @@ kubeadm_token_ttl: 0
+
+## Encrypting Secret Data at Rest
+encrypt_secret_data: false
+encrypt_token: "{{ lookup('password', 'credentials/encrypt_token length=32 chars=ascii_letters,digits') }}"
```
This was recently changed so that it went inside inventory dir:
```encrypt_token: "{{ lookup('password', inventory_dir + 'credentials/encrypt_token length=32 chars=ascii_letters,digits') }}" 
```

woopstar replied:
I dont get this comment?

mattymo commented on roles/kubernetes/master/defaults/main.yml:
```diff
@@ -88,3 +88,8 @@ kubeadm_token_ttl: 0
 volume_cross_zone_attachment: false
+
+## Encrypting Secret Data at Rest
+encrypt_secret_data: false
```
Please prefix your variables with `kube_`
These should be kube_encrypt_secret_data, kube_encrypt_token, and kube_encryption_algorithm.

mattymo commented on roles/kubernetes/master/tasks/encrypt-at-rest.yml:
```diff
@@ -0,0 +1,8 @@
+    dest: "{{ kube_config_dir }}/secrets_encryption.yaml"
+    mode: 0640
+  tags:
+    - kube-apiserver
```
missing condition:
`when: kube_encrypt_secret_data`

woopstar replied:
The whole file is not run if `kube_encrypt_secret_data` is not set. See:

```
- import_tasks: encrypt-at-rest.yml
  when: encrypt_secret_data
```

mattymo commented on roles/kubernetes/master/templates/kubeadm-config.yaml.j2:
```diff
@@ -48,6 +48,9 @@ apiServerExtraArgs:
   oidc-groups-claim: {{ kube_oidc_groups_claim }}
 {%   endif %}
+{% endif %}
+{% if encrypt_secret_data|default(false) %}
```
remove default, change to {% if kube_encrypt_secret_data %}

mattymo commented on roles/kubernetes/master/templates/manifests/kube-apiserver.manifest.j2:
```diff
@@ -97,6 +97,9 @@ spec:
 {% if authorization_modes %}
     - --authorization-mode={{ authorization_modes|join(',') }}
 {% endif %}
+{% if encrypt_secret_data|default(false) %}
```
ditto

woopstar commented:
@mattymo can we merge this?

mattymo commented on roles/kubernetes/master/defaults/main.yml:
```diff
@@ -88,3 +88,8 @@ kubeadm_token_ttl: 0
+
+## Encrypting Secret Data at Rest
+kube_encrypt_secret_data: false
+kube_encrypt_token: "{{ lookup('password', 'credentials/kube_encrypt_token length=32 chars=ascii_letters,digits') }}"
```
The credentials dir was moved recently. Change this line to
```
kube_encrypt_token: "{{ lookup('password', inventory_dir + '/credentials/kube_encrypt_token length=32 chars=ascii_letters,digits') }}"
```

bradbeam commented:
can you rebase your changes off master? i think that should fix up ci

woopstar commented:
Rebased it this way:

```
$ git remote add kubespray https://github.com/kubernetes-incubator/kubespray

$ git fetch --all
Fetching origin
Fetching kubespray
remote: Counting objects: 240, done.
remote: Compressing objects: 100% (28/28), done.
remote: Total 240 (delta 105), reused 98 (delta 95), pack-reused 115
Receiving objects: 100% (240/240), 55.07 KiB | 18.36 MiB/s, done.
Resolving deltas: 100% (127/127), completed with 56 local objects.
From https://github.com/kubernetes-incubator/kubespray
 * [new branch]        fix_bool_assert                        -> kubespray/fix_bool_assert
 * [new branch]        master                                 -> kubespray/master
 * [new branch]        numbering_includes                     -> kubespray/numbering_includes
 * [new branch]        preempt_ci_instances                   -> kubespray/preempt_ci_instances
 * [new branch]        releasetest                            -> kubespray/releasetest
 * [new branch]        releasetest-1                          -> kubespray/releasetest-1
 * [new branch]        revert-1778-etcd_sec_fixes             -> kubespray/revert-1778-etcd_sec_fixes
 * [new branch]        revert-1809-revert-1795-update_roadmap -> kubespray/revert-1809-revert-1795-update_roadmap
 * [new branch]        revert-1937-disable-api-insecure-port  -> kubespray/revert-1937-disable-api-insecure-port
 * [new branch]        rkt_support                            -> kubespray/rkt_support
 * [new branch]        rpm_specs                              -> kubespray/rpm_specs
 * [new branch]        terraform-aws-inventory                -> kubespray/terraform-aws-inventory
 * [new branch]        testpr2                                -> kubespray/testpr2
 * [new tag]           v2.4.0                                 -> v2.4.0

$ git rebase -i kubespray/master
Successfully rebased and updated refs/heads/encrypt-secrets-patch.

$ git pull
Merge made by the 'recursive' strategy.

$ git push origin encrypt-secrets-patch
Everything up-to-date
```

gopinatht commented:
Hey guys, what are the chances of this making the next release and when is the next release?

woopstar commented:
New PR in #2428 - Rebase here is messed up
